File: src/core/secrets.js
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import TOML from '@iarna/toml';

const RULES_PATH = fileURLToPath(new URL('../../vendor/gitleaks.toml', import.meta.url));

let compiledRules = null;

function loadRules() {
  if (compiledRules) return compiledRules;
  const raw = readFileSync(RULES_PATH, 'utf8');
  const parsed = TOML.parse(raw);
  const ruleSet = (parsed.rules || []).map((r) => {
    // Skip path-only rules (no regex field)
    if (!r.regex) return null;
    try {
      const allowPatterns = (r.allowlists || []).flatMap((al) =>
        (al.regexes || []).map((rx) => new RegExp(rx))
      );
      return {
        id: r.id,
        kind: r.id,
        re: new RegExp(r.regex, 'g'),
        keywords: r.keywords || [],
        allowPatterns,
      };
    } catch {
      return null;
    }
  }).filter(Boolean);

  ruleSet.push({
    id: 'jwt',
    kind: 'jwt',
    re: /eyJ[A-Za-z0-9_-]+\.eyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+/g,
    keywords: [],
    allowPatterns: [],
  });

  compiledRules = ruleSet;
  return ruleSet;
}

export function detectSecret(text) {
  if (!text) return null;
  const rules = loadRules();
  for (const rule of rules) {
    rule.re.lastIndex = 0;
    const m = rule.re.exec(text);
    if (m) {
      const value = m[0];
      // Check allowlists — if any allowlist pattern matches, skip this hit
      const allowed = rule.allowPatterns.some((ap) => ap.test(value));
      if (allowed) continue;
      return {
        kind: rule.kind,
        value,
        lastChars: value.slice(-Math.min(6, value.length)),
        startIndex: m.index,
        endIndex: m.index + value.length,
      };
    }
  }

  const ccMatch = text.match(/\b(?:\d[ -]?){13,19}\b/);
  if (ccMatch) {
    const digits = ccMatch[0].replace(/[ -]/g, '');
    if (digits.length >= 13 && digits.length <= 19 && luhnValid(digits)) {
      return {
        kind: 'credit_card',
        value: digits,
        lastChars: digits.slice(-4),
        startIndex: ccMatch.index,
        endIndex: ccMatch.index + ccMatch[0].length,
      };
    }
  }

  return null;
}

function luhnValid(digits) {
  let sum = 0;
  let alt = false;
  for (let i = digits.length - 1; i >= 0; i--) {
    let n = parseInt(digits[i], 10);
    if (alt) {
      n *= 2;
      if (n > 9) n -= 9;
    }
    sum += n;
    alt = !alt;
  }
  return sum % 10 === 0;
}

File: src/core/secrets.rs
use once_cell::sync::Lazy;
use regex::Regex;
use serde::Deserialize;
use std::sync::Mutex;

#[derive(Debug)]
pub struct SecretHit {
    pub kind: String,
    pub value: String,
    pub last_chars: String,
}

#[derive(Debug, Deserialize)]
struct GitleaksConfig {
    rules: Option<Vec<RawRule>>,
}

#[derive(Debug, Deserialize)]
struct RawRule {
    id: String,
    regex: Option<String>,
}

struct CompiledRule { kind: String, re: Regex }

static RULES: Lazy<Mutex<Vec<CompiledRule>>> = Lazy::new(|| {
    let raw = include_str!("../../vendor/gitleaks.toml");
    let cfg: GitleaksConfig = toml::from_str(raw).unwrap_or(GitleaksConfig { rules: None });
    let mut out: Vec<CompiledRule> = cfg.rules.unwrap_or_default().into_iter()
        .filter_map(|r| {
            let pattern = r.regex.as_deref()?;
            let re = Regex::new(pattern).ok()?;
            Some(CompiledRule { kind: r.id, re })
        })
        .collect();
    out.push(CompiledRule {
        kind: "jwt".into(),
        re: Regex::new(r"eyJ[A-Za-z0-9_-]+\.eyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+").unwrap(),
    });
    Mutex::new(out)
});

pub fn detect_secret(text: &str) -> Option<SecretHit> {
    if text.is_empty() { return None; }
    let rules = RULES.lock().ok()?;
    for rule in rules.iter() {
        if let Some(m) = rule.re.find(text) {
            let value = m.as_str().to_string();
            let last_chars = value.chars().rev().take(6).collect::<String>().chars().rev().collect();
            return Some(SecretHit { kind: rule.kind.clone(), value, last_chars });
        }
    }
    drop(rules);
    detect_credit_card(text)
}

fn detect_credit_card(text: &str) -> Option<SecretHit> {
    let cc_re = Regex::new(r"\b(?:\d[ -]?){13,19}\b").ok()?;
    let m = cc_re.find(text)?;
    let raw = m.as_str();
    let digits: String = raw.chars().filter(|c| c.is_ascii_digit()).collect();
    if !(13..=19).contains(&digits.len()) || !luhn(&digits) { return None; }
    let last_chars = digits.chars().rev().take(4).collect::<String>().chars().rev().collect();
    Some(SecretHit { kind: "credit_card".into(), value: digits, last_chars })
}

fn luhn(s: &str) -> bool {
    let mut sum = 0u32;
    let mut alt = false;
    for c in s.chars().rev() {
        let mut n = c.to_digit(10).unwrap_or(0);
        if alt { n *= 2; if n > 9 { n -= 9; } }
        sum += n;
        alt = !alt;
    }
    sum % 10 == 0
}
